src/pyqasm/visitor.py

```diff
@@ -559,7 +576,54 @@ def _check_variable_cast_type(
                 span=statement.span,
             )
 
-    def _visit_measurement(  # pylint: disable=too-many-locals
+    def _qubit_register_consolidation(
+        self, unrolled_stmts: list, total_qubits: int
+    ) -> list[qasm3_ast.Statement]:
+        """
+        Consolidate all quantum registers into a single register '__PYQASM_QUBITS__'.
+
+        Args:
+            unrolled_stmts (list): The list of statements to process and modify in-place.
+
+        Raises:
+            ValidationError: If the total number of qubits exceeds the available device qubits,
+                             or if the reserved register '__PYQASM_QUBITS__' is already declared
+                             in the original QASM program.
+        """
+        if total_qubits > self._device_qubits:  # type: ignore
+            raise_qasm3_error(
+                f"Total qubits '({total_qubits})' exceed device qubits '({self._device_qubits})'.",
+            )
+
+        if "__PYQASM_QUBITS__" in self._global_qreg_size_map:
+            raise_qasm3_error(
+                "Original QASM program already declares quantum register '__PYQASM_QUBITS__'.",
+            )
+
+        if "__PYQASM_QUBITS__" in self._global_creg_size_map:
+            raise_qasm3_error(
+                "Original QASM program already declares classical register '__PYQASM_QUBITS__'.",
+            )
+
+        global_scope = self._get_global_scope()
+        if "__PYQASM_QUBITS__" in global_scope:
+            raise_qasm3_error(
+                "Variable '__PYQASM_QUBITS__' is already exists",
```

Comment:
```suggestion
                "Variable '__PYQASM_QUBITS__' is already defined",
```

Please also use the `span` of the variable in the error reporting

Reply:
at this level, we need extra code around the visitor, currently `Variable()`don't have span element. i will add that and search around the global scope variables and raise the error if found. 
Does that works..?